chainntnfs/bitcoindnotify/bitcoind.go

```diff
@@ -830,3 +870,137 @@ func (b *BitcoindNotifier) RegisterBlockEpochNtfn() (*chainntnfs.BlockEpochEvent
 		}, nil
 	}
 }
+
+// handleMissedBlocks is called when the chain backend misses a series of
+// blocks. It dispatches all relevant notifications for the missed blocks,
+// handling a reorg if necessary.
+func (b *BitcoindNotifier) handleMissedBlocks(newHeight int32) error {
+	// If a reorg causes our best hash to be incorrect, rewind the chain
+	// so our best block is set to the closest common ancestor, then dispatch
+	// notifications from there
+	hashAtBestHeight, err := b.chainConn.GetBlockHash(int64(b.bestBlock.Height))
+	if err != nil {
+		return fmt.Errorf("unable to find blockhash for height=%d: %v",
+			b.bestBlock.Height, err)
+	}
+	commonAncestorHeight, err := b.getCommonBlockAncestorHeight(
+		*b.bestBlock.Hash,
+		*hashAtBestHeight,
+	)
+	if err != nil {
+		return fmt.Errorf("unable to find common ancestor: %v", err)
+	}
+	if err := b.rewindChain(commonAncestorHeight); err != nil {
+		return fmt.Errorf("unable to rewind chain: %v", err)
+	}
+
+	// We want to start dispatching historical notifications from the block
+	// right after our best block, to avoid a redundant notification.
+	for height := commonAncestorHeight + 1; height < newHeight; height++ {
+		hash, err := b.chainConn.GetBlockHash(int64(height))
+		if err != nil {
+			return fmt.Errorf("unable to get block hash: %v", err)
+		}
+		chainntnfs.Log.Infof("Catching up on missed block w/ height %d", height)
+		if err := b.handleBlockConnected(height, *hash); err != nil {
+			return fmt.Errorf("error on handling connected block: %v",
+				err)
+		}
+	}
+	return nil
+}
+
+// rewindChain handles internal state updates associated with disconnected blocks.
+// It has no effect if given a height greater than or equal to our current best
+// known height.
+func (b *BitcoindNotifier) rewindChain(newBestHeight int32) error {
```

Comment:
looks like also can be extracted into a common method that takes the `chainConn`?